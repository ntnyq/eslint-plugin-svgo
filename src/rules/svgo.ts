import { join } from 'node:path'
import { messages, reportDifferences } from 'eslint-formatting-reporter'
import { createSyncFn } from 'synckit'
import { dirWorkers } from '../dir'
import { svgoConfigProperties } from '../schema'
import { booleanSchema, stringSchema } from '../schema/shared'
import type { Rule } from 'eslint'
import type { Config } from 'svgo'
import type { SvgoParserError } from '../types'

// import from svgo next release
type Output = { data: string }

let optimizeSVG: (input: string, config: Config) => Output

export const svgo = {
  meta: {
    type: 'suggestion',
    docs: {
      recommended: true,
      description: 'Use SVGO to optimize SVG files',
      url: 'https://github.com/ntnyq/eslint-plugin-svgo',
    },
    fixable: 'code',
    schema: [
      {
        type: 'object',
        properties: {
          ...svgoConfigProperties,

          svgoConfig: {
            description: 'Use an external config file, e.g: svgo.config.mjs',
            oneOf: [booleanSchema, stringSchema],
          },
        },
        additionalProperties: false,
      },
    ],
    messages,
    defaultOptions: [
      {
        js2svg: {
          indent: 2,
          pretty: true,
        },
        plugins: ['preset-default'],
      },
    ],
  },
  create(context) {
    if (!optimizeSVG) {
      optimizeSVG = createSyncFn(join(dirWorkers, 'svgo.cjs')) as any
    }

    return {
      Program() {
        const sourceCode = context.sourceCode.text

        try {
          const output = optimizeSVG(sourceCode, {
            path: context.filename,
            /* v8 ignore next */
            ...(context.options?.[0] || {}),
          })

          reportDifferences(context, sourceCode, output.data)
        } catch (err) {
          if ((err as Error)?.name === 'SvgoParserError') {
            const { reason, line, column } = err as SvgoParserError

            context.report({
              message: `SvgoParserError: ${reason}`,
              loc: {
                start: { line, column },
                end: { line, column },
              },
            })
            /* v8 ignore start */
          } else {
            context.report({
              loc: {
                start: { line: 1, column: 0 },
                end: { line: 1, column: 0 },
              },
              message: 'Failed to optimize SVG file',
            })
          }
          /* v8 ignore end */
        }
      },
    }
  },
} satisfies Rule.RuleModule
