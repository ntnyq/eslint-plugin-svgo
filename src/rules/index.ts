import { svgo } from './svgo'

export const rules = {
  svgo,
}
