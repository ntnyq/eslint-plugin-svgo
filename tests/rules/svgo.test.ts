import { expect } from 'vitest'
import { resolve } from '../../scripts/utils'
import { svgo as rule } from '../../src/rules/svgo'
import { $, run } from '../internal'

run({
  name: 'svgo',
  rule,
  invalid: [
    /**
     * ===========================================
     *                 js2svg
     * ===========================================
     */
    {
      description: 'js2svg default',
      filename: 'file.svg',
      options: {
        plugins: [],
      },
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<?xml version="1.0" encoding="UTF-8"?>
          <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!--Created with love!-->
            <defs>
              <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
            </defs>
            <g>
              <use href="#circle" transform="skewX(16)"/>
              <rect id="useless" width="0" height="0" fill="#ff0000"/>
            </g>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'js2svg pretty false',
      filename: 'file.svg',
      options: {
        js2svg: {
          pretty: false,
        },
        plugins: [],
      },
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(
          `"<?xml version="1.0" encoding="UTF-8"?><svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150"><!--Created with love!--><defs><ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/></defs><g><use href="#circle" transform="skewX(16)"/><rect id="useless" width="0" height="0" fill="#ff0000"/></g></svg>"`,
        )
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'js2svg indent 4',
      filename: 'file.svg',
      options: {
        js2svg: {
          indent: 4,
        },
        plugins: [],
      },
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<?xml version="1.0" encoding="UTF-8"?>
          <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!--Created with love!-->
              <defs>
                  <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
              </defs>
              <g>
                  <use href="#circle" transform="skewX(16)"/>
                  <rect id="useless" width="0" height="0" fill="#ff0000"/>
              </g>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },

    /**
     * ===========================================
     *                 preset-default
     * ===========================================
     */
    {
      description: 'default preset-default',
      filename: 'file.svg',
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100">
            <defs>
              <circle id="a" cx="50" cy="50" r="50" fill="#000"/>
            </defs>
            <use href="#a" transform="skewX(16)"/>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'preset-default disable cleanupIds',
      filename: 'file.svg',
      options: {
        plugins: [
          {
            name: 'preset-default',
            params: {
              overrides: {
                cleanupIds: false,
              },
            },
          },
        ],
      },
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100">
            <defs>
              <circle id="circle" cx="50" cy="50" r="50" fill="#000"/>
            </defs>
            <use href="#circle" transform="skewX(16)"/>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'preset-default with cleanupIds params',
      filename: 'file.svg',
      options: {
        plugins: [
          {
            name: 'preset-default',
            params: {
              overrides: {
                cleanupIds: {
                  // minify referenced IDs
                  minify: false,
                },
              },
            },
          },
        ],
      },
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100">
            <defs>
              <circle id="circle" cx="50" cy="50" r="50" fill="#000"/>
            </defs>
            <use href="#circle" transform="skewX(16)"/>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },

    /**
     * ===========================================
     *                 plugins
     * ===========================================
     */
    {
      description: 'plugin cleanupIds name',
      filename: 'file.svg',
      options: {
        plugins: ['cleanupIds'],
      },
      code: $`
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 100" width="150">
          <rect id="useless" width="0" height="0" fill="#ff0000"/>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 100" width="150">
            <rect width="0" height="0" fill="#ff0000"/>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'plugin cleanupIds object',
      filename: 'file.svg',
      options: {
        plugins: [
          {
            name: 'cleanupIds',
          },
        ],
      },
      code: $`
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 100" width="150">
          <rect id="useless" width="0" height="0" fill="#ff0000"/>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 100" width="150">
            <rect width="0" height="0" fill="#ff0000"/>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'plugin cleanupIds with params',
      filename: 'file.svg',
      options: {
        plugins: [
          {
            name: 'cleanupIds',
            params: {
              remove: false,
            },
          },
        ],
      },
      code: $`
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 100" width="150">
          <rect id="useless" width="0" height="0" fill="#ff0000"/>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 100" width="150">
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },

    /**
     * ===========================================
     *                 svgoConfig
     * ===========================================
     */
    {
      description: 'svgoConfig default',
      filename: 'file.svg',
      options: {
        svgoConfig: true,
      },
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(
          `
          "<?xml version="1.0" encoding="UTF-8"?>
          <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!--Created with love!-->
              <defs>
                  <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="a"/>
              </defs>
              <g>
                  <use href="#a" transform="skewX(16)"/>
                  <rect width="0" height="0" fill="#ff0000"/>
              </g>
          </svg>
          "
        `,
        )
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'svgoConfig config path',
      filename: 'file.svg',
      options: {
        svgoConfig: resolve('svgo.config.mjs'),
      },
      code: $`
        <?xml version="1.0" encoding="UTF-8"?>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!-- Created with love! -->
          <defs>
            <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/>
          </defs>
          <g>
            <use href="#circle" transform="skewX(16)"/>
            <rect id="useless" width="0" height="0" fill="#ff0000"/>
          </g>
        </svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "<?xml version="1.0" encoding="UTF-8"?>
          <svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150">
          <!--Created with love!-->
              <defs>
                  <ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="a"/>
              </defs>
              <g>
                  <use href="#a" transform="skewX(16)"/>
                  <rect width="0" height="0" fill="#ff0000"/>
              </g>
          </svg>
          "
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },

    /**
     * ===========================================
     *                parser error
     * ===========================================
     */
    {
      description: 'invalid',
      filename: 'file.svg',
      code: $`
        <svg>
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`"<svg>"`)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      description: 'not svg',
      filename: 'file.svg',
      code: $`
        name: CI
        on:
          push:
            branches:
              - main
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`
          "name: CI
          on:
            push:
              branches:
                - main"
        `)
      },
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
  ],
})
