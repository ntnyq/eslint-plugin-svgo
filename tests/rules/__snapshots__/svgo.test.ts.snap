// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svgo > invalid > Invalid #0: js2svg default 1`] = `
[
  {
    "column": 39,
    "endColumn": 1,
    "endLine": 11,
    "fix": {
      "range": [
        38,
        367,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150"><!--Created with love!--><defs><ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/></defs><g><use href="#circle" transform="skewX(16)"/><rect id="useless" width="0" height="0" fill="#ff0000"/></g>",
    },
    "line": 1,
    "message": "Replace \`⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150">⏎··<!--·Created·with·love!·-->⏎··<defs>⏎····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/>⏎··</defs>⏎··<g>⏎····<use·href="#circle"·transform="skewX(16)"/>⏎····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150"><!--Created·with·love!--><defs><ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/></defs><g><use·href="#circle"·transform="skewX(16)"/><rect·id="useless"·width="0"·height="0"·fill="#ff0000"/></g>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #1: js2svg pretty false 1`] = `
[
  {
    "column": 39,
    "endColumn": 1,
    "endLine": 11,
    "fix": {
      "range": [
        38,
        367,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150"><!--Created with love!--><defs><ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/></defs><g><use href="#circle" transform="skewX(16)"/><rect id="useless" width="0" height="0" fill="#ff0000"/></g>",
    },
    "line": 1,
    "message": "Replace \`⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150">⏎··<!--·Created·with·love!·-->⏎··<defs>⏎····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/>⏎··</defs>⏎··<g>⏎····<use·href="#circle"·transform="skewX(16)"/>⏎····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150"><!--Created·with·love!--><defs><ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/></defs><g><use·href="#circle"·transform="skewX(16)"/><rect·id="useless"·width="0"·height="0"·fill="#ff0000"/></g>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #2: js2svg indent 4 1`] = `
[
  {
    "column": 39,
    "endColumn": 1,
    "endLine": 11,
    "fix": {
      "range": [
        38,
        367,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0  150 100 " width="150"><!--Created with love!--><defs><ellipse cx="50" cy="50.0" rx="50.00" ry="auto" fill="black" id="circle"/></defs><g><use href="#circle" transform="skewX(16)"/><rect id="useless" width="0" height="0" fill="#ff0000"/></g>",
    },
    "line": 1,
    "message": "Replace \`⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150">⏎··<!--·Created·with·love!·-->⏎··<defs>⏎····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/>⏎··</defs>⏎··<g>⏎····<use·href="#circle"·transform="skewX(16)"/>⏎····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150"><!--Created·with·love!--><defs><ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/></defs><g><use·href="#circle"·transform="skewX(16)"/><rect·id="useless"·width="0"·height="0"·fill="#ff0000"/></g>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #3: default preset-default 1`] = `
[
  {
    "column": 1,
    "endColumn": 1,
    "endLine": 11,
    "fix": {
      "range": [
        0,
        367,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100"><defs><circle id="a" cx="50" cy="50" r="50" fill="#000"/></defs><use href="#a" transform="skewX(16)"/>",
    },
    "line": 1,
    "message": "Replace \`<?xml·version="1.0"·encoding="UTF-8"?>⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150">⏎··<!--·Created·with·love!·-->⏎··<defs>⏎····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/>⏎··</defs>⏎··<g>⏎····<use·href="#circle"·transform="skewX(16)"/>⏎····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·width="150"·viewBox="0·0·150·100"><defs><circle·id="a"·cx="50"·cy="50"·r="50"·fill="#000"/></defs><use·href="#a"·transform="skewX(16)"/>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #4: preset-default disable cleanupIds 1`] = `
[
  {
    "column": 1,
    "endColumn": 1,
    "endLine": 11,
    "fix": {
      "range": [
        0,
        367,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100"><defs><circle id="circle" cx="50" cy="50" r="50" fill="#000"/></defs><use href="#circle" transform="skewX(16)"/>",
    },
    "line": 1,
    "message": "Replace \`<?xml·version="1.0"·encoding="UTF-8"?>⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150">⏎··<!--·Created·with·love!·-->⏎··<defs>⏎····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/>⏎··</defs>⏎··<g>⏎····<use·href="#circle"·transform="skewX(16)"/>⏎····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·width="150"·viewBox="0·0·150·100"><defs><circle·id="circle"·cx="50"·cy="50"·r="50"·fill="#000"/></defs><use·href="#circle"·transform="skewX(16)"/>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #5: preset-default with cleanupIds params 1`] = `
[
  {
    "column": 1,
    "endColumn": 1,
    "endLine": 11,
    "fix": {
      "range": [
        0,
        367,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100"><defs><circle id="circle" cx="50" cy="50" r="50" fill="#000"/></defs><use href="#circle" transform="skewX(16)"/>",
    },
    "line": 1,
    "message": "Replace \`<?xml·version="1.0"·encoding="UTF-8"?>⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150">⏎··<!--·Created·with·love!·-->⏎··<defs>⏎····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/>⏎··</defs>⏎··<g>⏎····<use·href="#circle"·transform="skewX(16)"/>⏎····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·width="150"·viewBox="0·0·150·100"><defs><circle·id="circle"·cx="50"·cy="50"·r="50"·fill="#000"/></defs><use·href="#circle"·transform="skewX(16)"/>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #6: plugin cleanupIds name 1`] = `
[
  {
    "column": 75,
    "endColumn": 1,
    "endLine": 3,
    "fix": {
      "range": [
        74,
        134,
      ],
      "text": "<rect width="0" height="0" fill="#ff0000"/>",
    },
    "line": 1,
    "message": "Replace \`⏎··<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎\` with \`<rect·width="0"·height="0"·fill="#ff0000"/>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #7: plugin cleanupIds object 1`] = `
[
  {
    "column": 75,
    "endColumn": 1,
    "endLine": 3,
    "fix": {
      "range": [
        74,
        134,
      ],
      "text": "<rect width="0" height="0" fill="#ff0000"/>",
    },
    "line": 1,
    "message": "Replace \`⏎··<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎\` with \`<rect·width="0"·height="0"·fill="#ff0000"/>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #8: plugin cleanupIds with params 1`] = `
[
  {
    "column": 75,
    "endColumn": 1,
    "endLine": 3,
    "fix": {
      "range": [
        74,
        134,
      ],
      "text": "<rect id="useless" width="0" height="0" fill="#ff0000"/>",
    },
    "line": 1,
    "message": "Replace \`⏎··<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎\` with \`<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #9: svgoConfig default 1`] = `
[
  {
    "column": 1,
    "endColumn": 76,
    "endLine": 2,
    "fix": {
      "range": [
        0,
        114,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100",
    },
    "line": 1,
    "message": "Replace \`<?xml·version="1.0"·encoding="UTF-8"?>⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·width="150"·viewBox="0·0·150·100\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 3,
    "endColumn": 1,
    "endLine": 4,
    "fix": {
      "range": [
        119,
        147,
      ],
      "text": "",
    },
    "line": 3,
    "message": "Delete \`<!--·Created·with·love!·-->⏎\`",
    "messageId": "delete",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 76,
    "endLine": 5,
    "fix": {
      "range": [
        156,
        231,
      ],
      "text": "        <circle id="a" cx="50" cy="50" r="50" fill="#000",
    },
    "line": 5,
    "message": "Replace \`····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle\` with \`········<circle·id="a"·cx="50"·cy="50"·r="50"·fill="#000\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 1,
    "endLine": 6,
    "fix": {
      "range": [
        235,
        235,
      ],
      "text": "  ",
    },
    "line": 6,
    "message": "Insert \`··\`",
    "messageId": "insert",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 3,
    "endColumn": 23,
    "endLine": 8,
    "fix": {
      "range": [
        247,
        273,
      ],
      "text": "  <use href="#a",
    },
    "line": 7,
    "message": "Replace \`<g>⏎····<use·href="#circle\` with \`··<use·href="#a\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 11,
    "fix": {
      "range": [
        299,
        373,
      ],
      "text": "</svg>
",
    },
    "line": 9,
    "message": "Replace \`····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎</svg>\` with \`</svg>⏎\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #10: svgoConfig config path 1`] = `
[
  {
    "column": 1,
    "endColumn": 76,
    "endLine": 2,
    "fix": {
      "range": [
        0,
        114,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100",
    },
    "line": 1,
    "message": "Replace \`<?xml·version="1.0"·encoding="UTF-8"?>⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·width="150"·viewBox="0·0·150·100\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 3,
    "endColumn": 1,
    "endLine": 4,
    "fix": {
      "range": [
        119,
        147,
      ],
      "text": "",
    },
    "line": 3,
    "message": "Delete \`<!--·Created·with·love!·-->⏎\`",
    "messageId": "delete",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 76,
    "endLine": 5,
    "fix": {
      "range": [
        156,
        231,
      ],
      "text": "        <circle id="a" cx="50" cy="50" r="50" fill="#000",
    },
    "line": 5,
    "message": "Replace \`····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle\` with \`········<circle·id="a"·cx="50"·cy="50"·r="50"·fill="#000\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 1,
    "endLine": 6,
    "fix": {
      "range": [
        235,
        235,
      ],
      "text": "  ",
    },
    "line": 6,
    "message": "Insert \`··\`",
    "messageId": "insert",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 3,
    "endColumn": 23,
    "endLine": 8,
    "fix": {
      "range": [
        247,
        273,
      ],
      "text": "  <use href="#a",
    },
    "line": 7,
    "message": "Replace \`<g>⏎····<use·href="#circle\` with \`··<use·href="#a\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 11,
    "fix": {
      "range": [
        299,
        373,
      ],
      "text": "</svg>
",
    },
    "line": 9,
    "message": "Replace \`····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎</svg>\` with \`</svg>⏎\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #11: invalid 1`] = `
[
  {
    "column": 6,
    "endColumn": 6,
    "endLine": 1,
    "line": 1,
    "message": "SvgoParserError: Unclosed root tag",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;

exports[`svgo > invalid > Invalid #12: not svg 1`] = `
[
  {
    "column": 2,
    "endColumn": 2,
    "endLine": 1,
    "line": 1,
    "message": "SvgoParserError: Non-whitespace before first tag.",
    "nodeType": null,
    "ruleId": "svgo",
    "severity": 2,
  },
]
`;
